src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +249,132 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
+                    return new SNITCPHandle(hostName, portNumber, timerExpire, callbackObject, parallel);
+                }
+                catch (Exception e)
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
+                    return null;
+                }
+            }
+            else
             {
                 SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                 return null;
             }
+        }
+
+        /// <summary>
+        /// Sends CLNT_UCAST_INST request for given instance name to SQL Sever Browser, and receive SVR_RESP from the Browser.
+        /// </summary>
+        /// <param name="browserEP">SQL Sever Browser IPEndPoint</param>
+        /// <param name="instanceName">instance name for CLNT_UCAST_INST request</param>
+        /// <returns>SVR_RESP packets from SQL Sever Browser</returns>
+        private static byte[] SendInstanceNameRequest(string browserHostname, string instanceName)
+        {
+            if (string.IsNullOrWhiteSpace(browserHostname) || string.IsNullOrWhiteSpace(instanceName))
+            {
+                throw new ArgumentException();
+            }
+
+            const byte CLNT_UCAST_INST = 0x04;
+            int byteCount = Encoding.ASCII.GetByteCount(instanceName);
+            byte[] requestPacket = new byte[byteCount + 1];
+            requestPacket[0] = CLNT_UCAST_INST;
+            Encoding.ASCII.GetBytes(instanceName, 0, instanceName.Length, requestPacket, 1);
 
-            return new SNITCPHandle(serverAndPortParts[0], portNumber, timerExpire, callbackObject, parallel);
+            byte[] responsePacket = null;
+            using (UdpClient client = new UdpClient(AddressFamily.InterNetwork))
+            {
+                Task<int> sendTask = client.SendAsync(requestPacket, requestPacket.Length, browserHostname, SqlServerBrowserPort);
+                Task<UdpReceiveResult> receiveTask = null;
+                if (sendTask.Wait(1000) && (receiveTask = client.ReceiveAsync()).Wait(1000))
+                {
+                    responsePacket = receiveTask.Result.Buffer;
+                }
+            }
+
+            return responsePacket;
+        }
+
+        /// <summary>
+        /// Finds port number for given instance name.
+        /// </summary>
+        /// <param name="browserHostname">SQL Sever Browser hostname</param>
+        /// <param name="instanceName">instance name to find port number</param>
+        /// <returns>port number for given instance name</returns>
+        private static int GetPortByInstanceName(string browserHostname, string instanceName)
+        {
+            if (string.IsNullOrWhiteSpace(browserHostname) || string.IsNullOrWhiteSpace(instanceName))
+            {
+                throw new ArgumentException();
```

Comment:
Seems like this should be an ArgumentNullException in the case where the argument is null rather than whitespace, and in either case the exception should include the name of the argument that was problematic?